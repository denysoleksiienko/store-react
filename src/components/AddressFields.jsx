import React from 'react';
import Form from 'react-bootstrap/Form';
import Col from 'react-bootstrap/Col';
import CountrySelect from 'react-bootstrap-country-select';

import { Label, InputsWrap, FormGroupInner } from 'styled';

export const AddressFields = ({ user, setAddress, setOptional, setCity, setCountry, setZip }) => (
  <InputsWrap>
    <Form.Group as={Col} md="10">
      <Label>Address</Label>
      <Form.Control
        type="text"
        placeholder="Street Address"
        value={user.address}
        onChange={(event) => setAddress(event.target.value)}
      />
    </Form.Group>
    <Form.Group as={Col} md="10">
      <Form.Control
        type="text"
        placeholder="Apt, Suite, Bldg, Gate Code. (optional)"
        value={user.optional}
        onChange={(event) => setOptional(event.target.value)}
      />
    </Form.Group>

    <Form.Group as={Col} md="10">
      <Form.Control
        type="text"
        placeholder="City"
        value={user.city}
        onChange={(event) => setCity(event.target.value)}
        required
      />
    </Form.Group>

    <FormGroupInner>
      <Form.Group as={Col} md="6">
        <CountrySelect value={user.country} onChange={(id) => setCountry(id)} placeholder="Country" flags={false} />
      </Form.Group>

      <Form.Group as={Col} md="4">
        <Form.Control
          type="text"
          placeholder="Zip"
          value={user.zip}
          onChange={(event) => setZip(event.target.value)}
          required
        />
      </Form.Group>
    </FormGroupInner>
  </InputsWrap>
);
