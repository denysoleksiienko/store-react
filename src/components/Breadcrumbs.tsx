import React from 'react';
import Breadcrumb from 'react-bootstrap/Breadcrumb';
import { BreadcrumbWrap, NavLinkStyled } from 'styled/Breadcrumbs';

export const Breadcrumbs: React.FC = () => (
  <BreadcrumbWrap>
    <Breadcrumb.Item linkAs={NavLinkStyled} linkProps={{ to: '/', exact: true }}>
      Shipping
    </Breadcrumb.Item>
    <Breadcrumb.Item linkAs={NavLinkStyled} linkProps={{ to: '/billing' }}>
      Billing
    </Breadcrumb.Item>
    <Breadcrumb.Item linkAs={NavLinkStyled} linkProps={{ to: '/payment' }}>
      Payment
    </Breadcrumb.Item>
  </BreadcrumbWrap>
);
