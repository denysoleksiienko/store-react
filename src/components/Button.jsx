import { StyledButton } from 'styled';

export const Button = ({ title, onClick }) => (
  <StyledButton variant="primary" size="lg" onClick={onClick} type="submit">
    {title}
  </StyledButton>
);
