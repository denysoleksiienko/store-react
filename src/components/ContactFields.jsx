import { useLocation } from 'react-router-dom';
import Col from 'react-bootstrap/Col';
import Form from 'react-bootstrap/Form';
import { Label, InputsWrap, InputSpan, FormGroupInner } from 'styled';

export const ContactFields = ({ user, setName, setPhone, setEmail }) => {
  const location = useLocation();

  return (
    <>
      {location.pathname === '/' ? (
        <InputsWrap>
          <Form.Group as={Col} md="10">
            <Label>Recipient</Label>
            <Form.Control
              type="text"
              placeholder="Full Name"
              value={user.name}
              onChange={(event) => setName(event.target.value)}
            />
          </Form.Group>
          <Form.Group as={Col} md="8">
            <FormGroupInner>
              <Form.Control
                type="phone"
                placeholder="Daytime Phone"
                value={user.phone}
                onChange={(event) => setPhone(event.target.value)}
              />
              <InputSpan>For delivery questions only</InputSpan>
            </FormGroupInner>
          </Form.Group>
        </InputsWrap>
      ) : (
        <InputsWrap>
          <Form.Group as={Col} md="10">
            <Label>Billing Contact</Label>
            <Form.Control
              type="text"
              placeholder="Full Name"
              value={user.name}
              onChange={(event) => setName(event.target.value)}
            />
          </Form.Group>
          <Form.Group as={Col} md="10">
            <Form.Control
              type="text"
              placeholder="Email Address"
              value={user.email}
              onChange={(event) => setEmail(event.target.value)}
            />
          </Form.Group>
        </InputsWrap>
      )}
    </>
  );
};
