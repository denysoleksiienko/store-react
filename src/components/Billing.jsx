import { useHistory } from 'react-router-dom';
import { useFormik } from 'formik';

import Col from 'react-bootstrap/Col';
import Form from 'react-bootstrap/Form';

import { Typeahead } from 'react-bootstrap-typeahead';
import { COUNTRIES } from 'constants/countries';

import { Button, Breadcrumbs } from 'components';
import { PAYMENT } from 'constants/pathnames';
import { Title, Label, FormGroupInner } from 'styled';

import { VALIDATION_BILLING } from 'constants/validationSchema';

export const Billing = (props) => {
  const history = useHistory();

  const formik = useFormik({
    initialValues: {
      ...props.user,
    },
    validationSchema: VALIDATION_BILLING,
    onSubmit: () => {
      setTimeout(() => history.push(PAYMENT), 300);
    },
  });

  return (
    <Col lg={7}>
      <Breadcrumbs />

      <Title>Billing Information</Title>

      <Form noValidate onSubmit={formik.handleSubmit}>
        <Form.Group as={Col} md="10">
          <Label>Billing Contact</Label>
          <Form.Control
            name="name"
            type="text"
            placeholder="Full Name"
            value={formik.values.name}
            onChange={(e) => {
              props.setName(e.target.value);
              formik.handleChange(e);
            }}
            onBlur={formik.handleBlur}
            isValid={formik.touched.name && !formik.errors.name}
            isInvalid={formik.touched.name && formik.errors.name}
          />
          <Form.Control.Feedback type="invalid">{formik.errors.name}</Form.Control.Feedback>
        </Form.Group>

        <Form.Group as={Col} md="10">
          <Form.Control
            name="email"
            type="email"
            placeholder="Email Address"
            value={props.user.email}
            onChange={(e) => {
              formik.handleChange(e);
              props.setEmail(e.target.value);
            }}
            onBlur={formik.handleBlur}
            isValid={formik.touched.email && formik.errors.email}
            isInvalid={formik.touched.email && formik.errors.email}
          />
          <Form.Control.Feedback type="invalid">{formik.errors.email}</Form.Control.Feedback>
        </Form.Group>

        <Form.Group as={Col} md="10">
          <Label>Billing Address</Label>

          <Form.Control
            name="address"
            type="text"
            placeholder="Street Address"
            value={formik.values.address}
            onChange={(e) => {
              props.setAddress(e.target.value);
              formik.handleChange(e);
            }}
            onBlur={formik.handleBlur}
            isValid={formik.touched.address && !formik.errors.address}
            isInvalid={formik.touched.address && formik.errors.address}
          />
          <Form.Control.Feedback type="invalid">{formik.errors.address}</Form.Control.Feedback>
        </Form.Group>
        <Form.Group as={Col} md="10">
          <Form.Control
            name="optional"
            type="text"
            placeholder="Apt, Suite, Bldg, Gate Code. (optional)"
            value={formik.values.optional}
            onChange={(e) => {
              props.setOptional(e.target.value);
              formik.handleChange(e);
            }}
            onBlur={formik.handleBlur}
            isValid={formik.touched.optional && !formik.errors.optional}
            isInvalid={formik.touched.optional && formik.errors.optional}
          />
          <Form.Control.Feedback type="invalid">{formik.errors.optional}</Form.Control.Feedback>
        </Form.Group>
        <Form.Group as={Col} md="10">
          <Form.Control
            name="city"
            type="text"
            placeholder="City"
            value={formik.values.city}
            onChange={(e) => {
              props.setCity(e.target.value);
              formik.handleChange(e);
            }}
            onBlur={formik.handleBlur}
            isValid={formik.touched.city && !formik.errors.city}
            isInvalid={formik.touched.city && formik.errors.city}
          />
          <Form.Control.Feedback type="invalid">{formik.errors.city}</Form.Control.Feedback>
        </Form.Group>
        <FormGroupInner>
          <Form.Group as={Col} md="6">
            <Typeahead
              id="basic-typeahead-single"
              placeholder="Country"
              onChange={(e) => props.setCountry(e)}
              options={COUNTRIES}
              selected={props.user.country}
              isValid={formik.touched.country && !formik.errors.country}
              isInvalid={formik.touched.country && formik.errors.country}
            />
            <Form.Control.Feedback type="invalid">{formik.errors.country}</Form.Control.Feedback>
          </Form.Group>

          <Form.Group as={Col} md="4">
            <Form.Control
              name="zip"
              type="text"
              placeholder="Zip"
              value={formik.values.zip}
              onChange={(e) => {
                props.setZip(e.target.value);
                formik.handleChange(e);
              }}
              // onBlur={formik.handleBlur}
              isValid={formik.touched.zip && !formik.errors.zip}
              isInvalid={formik.touched.zip && formik.errors.zip}
            />
            <Form.Control.Feedback type="invalid">{formik.errors.zip}</Form.Control.Feedback>
          </Form.Group>
        </FormGroupInner>

        <Form.Group as={Col} md="6">
          <Button title="Continue" />
        </Form.Group>
      </Form>
    </Col>
  );
};
