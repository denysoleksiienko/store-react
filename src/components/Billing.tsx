import React from 'react';
import { useHistory } from 'react-router-dom';
import { useFormik, FormikProps } from 'formik';

import Col from 'react-bootstrap/Col';
import Form from 'react-bootstrap/Form';

import { Button, Breadcrumbs, CountrySelect } from 'components';

import { COUNTRIES } from 'constants/countries';
import { PAYMENT } from 'constants/pathnames';

import { Title, Label, FormGroupInner, Error } from 'styled';

import { VALIDATION_BILLING } from 'constants/validationSchema';

import { IUserFormValues } from 'interfaces/FormValues';
import { IUserProps } from 'interfaces/UserProps';
import { ICountries } from 'interfaces/Countries';

export const Billing: React.FC<IUserProps> = ({
  user,
  setName,
  setEmail,
  setAddress,
  setOptional,
  setCity,
  setCountry,
  setZip,
}) => {
  const history = useHistory();

  const handleChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    switch (e.target.name) {
      case 'name':
        setName(e.target.value);
        break;
      case 'email':
        setEmail(e.target.value);
        break;
      case 'address':
        setAddress(e.target.value);
        break;
      case 'optional':
        setOptional(e.target.value);
        break;
      case 'city':
        setCity(e.target.value);
        break;
      case 'zip':
        setZip(e.target.value);
        break;
      default:
        e.target.value = '';
    }
    formik.handleChange(e);
  };

  const handleChangeCountry = (e: ICountries[]): void => {
    setCountry(e);
    formik.setFieldValue('country', e);
  };

  const formik: FormikProps<IUserFormValues> = useFormik({
    initialValues: { ...user },
    validationSchema: VALIDATION_BILLING,
    onSubmit: () => {
      setTimeout(() => history.push(PAYMENT), 300);
    },
  });

  return (
    <Col lg={7}>
      <Breadcrumbs />

      <Title>Billing Information</Title>

      <Form data-testid="form-billing" noValidate onSubmit={formik.handleSubmit}>
        <Form.Group as={Col} md="10">
          <Label>Billing Contact</Label>
          <Form.Control
            data-testid="name"
            name="name"
            type="text"
            placeholder="Full Name"
            value={formik.values.name}
            onChange={handleChange}
            onBlur={formik.handleBlur}
            isValid={formik.touched.name && !formik.errors.name}
            isInvalid={!!formik.errors.name && formik.touched.name}
            required
          />
          <Form.Control.Feedback type="invalid">{formik.errors.name}</Form.Control.Feedback>
        </Form.Group>

        <Form.Group as={Col} md="10">
          <Form.Control
            data-testid="email"
            name="email"
            type="email"
            placeholder="Email Address"
            value={user.email}
            onChange={handleChange}
            onBlur={formik.handleBlur}
            isValid={formik.touched.email && !formik.errors.email}
            isInvalid={!!formik.errors.email && formik.touched.email}
            required
          />
          <Form.Control.Feedback type="invalid">{formik.errors.email}</Form.Control.Feedback>
        </Form.Group>

        <Form.Group as={Col} md="10">
          <Label>Billing Address</Label>

          <Form.Control
            data-testid="address"
            name="address"
            type="text"
            placeholder="Street Address"
            value={formik.values.address}
            onChange={handleChange}
            onBlur={formik.handleBlur}
            isValid={formik.touched.address && !formik.errors.address}
            isInvalid={!!formik.errors.address && formik.touched.address}
            required
          />
          <Form.Control.Feedback type="invalid">{formik.errors.address}</Form.Control.Feedback>
        </Form.Group>
        <Form.Group as={Col} md="10">
          <Form.Control
            data-testid="optional"
            name="optional"
            type="text"
            placeholder="Apt, Suite, Bldg, Gate Code. (optional)"
            value={formik.values.optional}
            onChange={handleChange}
            onBlur={formik.handleBlur}
            isValid={formik.touched.optional && !formik.errors.optional}
            isInvalid={!!formik.errors.optional && formik.touched.optional}
            required
          />
          <Form.Control.Feedback type="invalid">{formik.errors.optional}</Form.Control.Feedback>
        </Form.Group>
        <Form.Group as={Col} md="10">
          <Form.Control
            data-testid="city"
            name="city"
            type="text"
            placeholder="City"
            value={formik.values.city}
            onChange={handleChange}
            onBlur={formik.handleBlur}
            isValid={formik.touched.city && !formik.errors.city}
            isInvalid={!!formik.errors.city && formik.touched.city}
            required
          />
          <Form.Control.Feedback type="invalid">{formik.errors.city}</Form.Control.Feedback>
        </Form.Group>
        <FormGroupInner>
          <Form.Group as={Col} md="6">
            <CountrySelect
              data-testid="country"
              placeholder="Country"
              onChange={handleChangeCountry}
              options={COUNTRIES}
              selected={formik.values.country}
              isValid={formik.touched.country && !formik.errors.country}
              isInvalid={!!formik.errors.country}
            />
            <Error>{formik.errors.country}</Error>
          </Form.Group>

          <Form.Group as={Col} md="4">
            <Form.Control
              data-testid="zip"
              name="zip"
              type="text"
              placeholder="Zip"
              value={formik.values.zip}
              onChange={handleChange}
              onBlur={formik.handleBlur}
              isValid={formik.touched.zip && !formik.errors.zip}
              isInvalid={!!formik.errors.zip && formik.touched.zip}
              required
            />
            <Form.Control.Feedback type="invalid">{formik.errors.zip}</Form.Control.Feedback>
          </Form.Group>
        </FormGroupInner>

        <Form.Group as={Col} md="6">
          <Button title="Continue" type="submit" />
        </Form.Group>
      </Form>
    </Col>
  );
};
