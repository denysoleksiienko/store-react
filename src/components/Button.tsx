import React from 'react';
import { StyledButton } from 'styled';

import { IButtonProps } from 'interfaces/Button';

export const Button: React.FC<IButtonProps> = ({ title, type }) => (
  <StyledButton variant="primary" size="lg" type={type}>
    {title}
  </StyledButton>
);
