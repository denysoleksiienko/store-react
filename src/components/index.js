import { connect } from 'react-redux';
import { fetchProducts } from 'redux/ducks/products';
import { setName, setPhone, setEmail, setAddress, setOptional, setCity, setCountry, setZip } from 'redux/ducks/user';
import { userState } from 'redux/selectors';

import { AddressFields as AddressFieldsComponent } from './AddressFields';
// import { ContactFields as ContactFieldsComponent } from './ContactFields';
import { Orders as OrdersComponent } from './Orders';
import { FinishOrders as FinishOrdersComponent } from './FinishOrders';

import { Shipping as ShippingComponent } from './Shipping';
import { Billing as BillingComponent } from './Billing';

export { Breadcrumbs } from './Breadcrumbs';
export { Button } from './Button';
export { Payment } from './Payment';
export { ContactFields } from './ContactFields';

const mapStateToProps = (state) => ({
  orders: state.orders.products,
  user: userState(state),
});

const mapDispatchToProps = {
  fetchProducts,
  setName,
  setPhone,
  setEmail,
  setAddress,
  setOptional,
  setCity,
  setCountry,
  setZip,
};

export const Orders = connect(mapStateToProps, mapDispatchToProps)(OrdersComponent);
export const Shipping = connect(mapStateToProps, mapDispatchToProps)(ShippingComponent);
export const Billing = connect(mapStateToProps, mapDispatchToProps)(BillingComponent);
export const AddressFields = connect(mapStateToProps, mapDispatchToProps)(AddressFieldsComponent);
// export const ContactFields = connect(mapStateToProps, mapDispatchToProps)(ContactFieldsComponent);
export const FinishOrders = connect(mapStateToProps, mapDispatchToProps)(FinishOrdersComponent);
