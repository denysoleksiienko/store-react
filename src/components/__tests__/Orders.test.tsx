import React from 'react';
import { getByLabelText, getByTestId, render, screen } from '@testing-library/react';
import renderer from 'react-test-renderer';

import { Provider } from 'react-redux';
import configureMockStore from 'redux-mock-store';

import { Orders } from 'components';
import { SummaryTitle, SummaryWrap, SummaryList } from 'styled';

const props = {
  orders: [],
  fetchProducts: jest.fn(),
};

const mockStore = configureMockStore();

describe('render Orders', () => {
  const store = mockStore({
    orders: {
      products: [
        {
          id: 1,
          img: '/img-bag.png',
          title: 'Bag',
          color: 'black',
          qty: 1,
          price: 50,
        },
      ],
      error: false,
    },
  });

  beforeEach(() => {
    render(
      <Provider store={store}>
        <Orders {...props} />
      </Provider>
    );
  });

  it('render Order Summary title', () => {
    expect(screen.getByText('Order Summary')).toBeInTheDocument();
  });

  it('render wrapper and list', () => {
    expect(getByTestId(document.documentElement, 'summ-wrap')).toBeInTheDocument();
    expect(getByTestId(document.documentElement, 'summ-list')).toBeInTheDocument();
  });

  it('should match snapshot', () => {
    const component = renderer
      .create(
        <Provider store={store}>
          <Orders {...props} />
        </Provider>
      )
      .toJSON();
    expect(component).toMatchSnapshot();
  });
});
