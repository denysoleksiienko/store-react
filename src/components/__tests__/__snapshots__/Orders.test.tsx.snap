// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render Orders should match snapshot 1`] = `
Array [
  <h5
    className="sc-kstrdz llctVS"
  >
    Order Summary
  </h5>,
  <div
    className="sc-hBEYos kpoNDI"
    data-testid="summ-wrap"
  >
    <div
      className="sc-fodVxV kZMtlF"
      data-testid="summ-list"
    >
      <div
        className="sc-bkzZxe gSnmCg"
      >
        <img
          alt="Bag"
          src="/img-bag.png"
        />
        <div
          className="sc-idOhPF tkVTk"
        >
          <span>
            Bag
          </span>
          <span>
            black
          </span>
          <span>
            Quantity: 
            1
          </span>
        </div>
        <span>
          $
          50
        </span>
      </div>
    </div>
    <div
      className="sc-fFubgz igwYDr"
    >
      <div
        className="sc-hHftDr gKZifq"
      >
        <span>
          Subtotal
        </span>
        <span>
          $
          50
        </span>
      </div>
      <div
        className="sc-hHftDr gKZifq"
      >
        <span>
          Shipping
        </span>
        <span>
          Free
        </span>
      </div>
      <div
        className="sc-hHftDr gKZifq"
      >
        <span>
          Taxes
        </span>
        <span>
          $10
        </span>
      </div>
      <hr />
      <div
        className="sc-hHftDr gKZifq"
      >
        <span
          className="sc-dIUggk jiCKLJ"
        >
          Total
        </span>
        <span
          className="sc-dIUggk jiCKLJ"
        >
          $
          60
        </span>
      </div>
    </div>
  </div>,
]
`;
